src/backend/joanie/core/api.py

```diff
@@ -127,6 +137,124 @@ def perform_create(self, serializer):
         owner = models.User.objects.get_or_create(username=username)[0]
         serializer.save(owner=owner)
 
+    @transaction.atomic
+    def create(self, request, *args, **kwargs):
+        product_id = request.data.get("product")
+        course_code = request.data.get("course")
+        billing_address = request.data.get("billing_address")
+        credit_card_id = request.data.get("credit_card_id")
+
+        try:
+            product = Product.objects.get(uid=product_id)
+        except Product.DoesNotExist:
+            return Response(f"Product {product_id} does not exist.", status=404)
+
+        # If product is not free, we have to create a payment.
+        # To create one, a billing address is mandatory
+        if product.price.amount > 0 and billing_address is None:
+            return Response({"billing_address": "This field is required."}, status=400)
+
+        # - Validate data then create an order
+        serializer = self.get_serializer(data=request.data)
```

Comment:
since you validate the data, this should be done first in the method and your lines 142 to 145 should be retrieved from serializer.validated_data. This way you can also let the serializer require the product field and retrieve the object.

Reply:
Oh yes it's clearly simpler to retrieve course and product objects.